File: examples/prime_sieve/ref.py
def sieve(n: int) -> int:
    primes = [True] * n

    # 0 and 1 are not primes
    if n > 0:
        primes[0] = False
    if n > 1:
        primes[1] = False

    p = 2
    while p * p < n:
        if primes[p]:
            j = p * p
            while j < n:
                primes[j] = False
                j += p
        p += 1

    count = 0
    i = 2
    while i < n:
        if primes[i]:
            count += 1
        i += 1

    return count


def main() -> int:
    limit = 50_000_000
    print("counting primes up to", limit)
    c = sieve(limit)
    print("answer:", c)
    return 0


if __name__ == "__main__":
    main()

File: examples/prime_sieve/ref.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int sieve(int n) {
    unsigned char *primes = malloc(n);
    if (!primes) {
        fprintf(stderr, "allocation failed\n");
        exit(1);
    }

    // initialize all entries to 1
    for (int i = 0; i < n; i++) {
        primes[i] = 1;
    }

    for (int p = 2; p * p < n; p++) {
        if (primes[p] == 1) {
            for (int j = p * p; j < n; j += p) {
                primes[j] = 0;
            }
        }
    }

    int count = 0;
    for (int i = 2; i < n; i++) {
        if (primes[i] == 1)
            count++;
    }

    free(primes);
    return count;
}

int main(void) {
    int limit = 50000000;
    printf("counting primes up to %d\n", limit);

    int c = sieve(limit);

    printf("answer: %d\n", c);
    return 0;
}
